### ISSUE TITLE
TypeError: extract_main() missing 1 required positional argument: 'checkpoint_folder'

### ISSUE THREAD

**Issue by DC95** [facebookresearch/vissl#362]
Hello @prigoyal and @QuentinDuval 

**What am I trying to do? -** 
I am trying to extract features from the VISSL trained Deepcluster_V2 model. I am using my own custom satellite data set, which is unlabelled, and I intend to study the features learned by the model using the test dataset for further analysis.

**What problem am I facing? -** 
While extracting the features I am facing this error - _**"TypeError: extract_main() missing 1 required positional argument: 'checkpoint_folder".**_ 

Some more information about the error is as follows - File "/p/home/jusers/chatterjee1/hdfml/operation_vissl/run_distributed_engines.py", line 166, in _distributed_worker
    process_main(cfg, dist_run_id, local_rank=local_rank, node_id=node_id)
    
**My understanding of the error -** 
I have checked line 166 of [run_distributed_engines.py](https://dl.fbaipublicfiles.com/vissl/tutorials/run_distributed_engines.py) and what I gather is that it is related to vissl.utils.checkpoint function **get_checkpoint_folder(config: AttrDict)**

The function cares about - config.CHECKPOINT.DIR (which I am already giving in my command)

I have followed the discussion in #235 and #292 and the stable version [tutorial](https://vissl.ai/tutorials/Feature_Extraction)

The command which I am using is - 

`python run_distributed_engines.py \hydra.verbose=true \config=pretrain/evaluate_features/simple_extract \+config/trunk_only=resnet50_layers \config.CHECKPOINT.DIR="/p/project/deepacf/kiste/DC/checkpoints_evaluate/" \config.MODEL.WEIGHTS_INIT.PARAMS_FILE="/p/project/deepacf/kiste/DC/checkpoints_tensorboard/model_final_checkpoint_phase149.torch"`

- The resnet50_layers.yaml is from vissl/configs/config/feature_extraction/trunk_only/rn50_layers.yaml

- Before you see how my simple_extract.yaml looks like, please note this is not my whole training configuration for the Deepcluster_V2 model. I collected information from _vissl/configs/config/feature_extraction/extract_resnet_in1k_8gpu.yaml_ and wrote this YAML file.

**simple_extract.yaml file -** 

> engine_name: extract_features
> config:
>   DATA:
>     NUM_DATALOADER_WORKERS: 5
>     TRAIN:
>       DATA_SOURCES: [disk_folder]
>       DATASET_NAMES: [dummy_data_folder]
>       BATCHSIZE_PER_REPLICA: 16
>       LABEL_TYPE: sample_index
>       TRANSFORMS:            
>         - name: ImgPilToMultiCrop
>           total_num_crops: 2
>           size_crops: [490]
>           num_crops: [2]
>           crop_scales: [[0.08, 1]]
>         - name: RandomHorizontalFlip
>           p: 0.5
>         - name: ImgPilColorDistortion
>           strength: 1.0
>         - name: ImgPilGaussianBlur
>           p: 0.5
>           radius_min: 0.1
>           radius_max: 2.0
>         - name: ToTensor
>         - name: Normalize
>           mean: [0.004819399821419931, 0.004819357153477664, 0.004819095912873002]
>           std: [0.05880587872490098, 0.05880556607209096, 0.05880256329021113]
>       COLLATE_FUNCTION: multicrop_collator #
>       MMAP_MODE: True
>       COPY_TO_LOCAL_DISK: False
>       COPY_DESTINATION_DIR: /tmp/imagenet1k/
>       DROP_LAST: True
>     TEST:
>       DATA_SOURCES: [disk_folder]
>       DATASET_NAMES: [dummy_data_folder]
>       BATCHSIZE_PER_REPLICA: 16
>       LABEL_TYPE: sample_index
>       TRANSFORMS:
>         - name: ImgPilToMultiCrop
>           total_num_crops: 2
>           size_crops: [490]
>           num_crops: [2]
>           crop_scales: [[0.08, 1]]
>         - name: RandomHorizontalFlip
>           p: 0.5
>         - name: ImgPilColorDistortion
>           strength: 1.0
>         - name: ImgPilGaussianBlur
>           p: 0.5
>           radius_min: 0.1
>           radius_max: 2.0
>         - name: ToTensor
>         - name: Normalize
>         - mean: [0.004819399821419931, 0.004819357153477664, 0.004819095912873002]
>           std: [0.05880587872490098, 0.05880556607209096, 0.05880256329021113]
>       COLLATE_FUNCTION: multicrop_collator #
>       MMAP_MODE: False
>       COPY_TO_LOCAL_DISK: False
>       COPY_DESTINATION_DIR: /tmp/imagenet1k/
>       ENABLE_QUEUE_DATASET: False
>   MODEL:
>     TRUNK:
>       NAME: resnet
>       RESNETS:
>         DEPTH: 50
>     WEIGHTS_INIT:
>       PARAMS_FILE: /p/project/deepacf/kiste/DC/checkpoints_tensorboard/model_final_checkpoint_phase149.torch
>       APPEND_PREFIX: trunk.base_model._feature_blocks.
>       STATE_DICT_KEY_NAME: classy_state_dict
>   DISTRIBUTED:
>     NUM_NODES: 1
>     NUM_PROC_PER_NODE: 4
>     RUN_ID: auto
>   MACHINE:
>     DEVICE: gpu
>   CHECKPOINT:
>     AUTO_RESUME: True
>     DIR: /p/project/deepacf/kiste/DC/checkpoints_evaluate

**For trying purpose**
For a trial, I thought maybe it needs the original training configuration along with freeze trunk info. I tried with the original training file configuration in place of the simple_extract.yaml. But it shows the same error.

The training file configuration used for feature extraction looks like this - 

> engine_name: extract_features
> config:
>   VERBOSE: True
>   LOG_FREQUENCY: 10
>   TEST_ONLY: False
>   TEST_MODEL: False
>   SEED_VALUE: 0
>   MULTI_PROCESSING_METHOD: forkserver
>   HOOKS:
>     PERF_STATS:
>       MONITOR_PERF_STATS: True 
>       ROLLING_BTIME_FREQ: 313  
>   DATA:
>     NUM_DATALOADER_WORKERS: 5   
>     TRAIN: 
>       DATA_SOURCES: [disk_folder] 
>       DATASET_NAMES: [dummy_data_folder]
>       BATCHSIZE_PER_REPLICA: 16  
>       LABEL_TYPE: sample_index    
>       TRANSFORMS:               
>         - name: ImgPilToMultiCrop
>           total_num_crops: 2
>           size_crops: [490]
>           num_crops: [2]
>           crop_scales: [[0.08, 1]]
>         - name: RandomHorizontalFlip 
>           p: 0.5
>         - name: ImgPilColorDistortion
>           strength: 1.0
>         - name: ImgPilGaussianBlur
>           p: 0.5
>           radius_min: 0.1
>           radius_max: 2.0
>         - name: ToTensor          # (H x W x C)[0, 255] to (C x H x W) [0.0, 1.0]
>         - name: Normalize
>           # mean: [0.485, 0.456, 0.406]
>           # std: [0.229, 0.224, 0.225]
>           mean: [0.004819399821419931, 0.004819357153477664, 0.004819095912873002]
>           std: [0.05880587872490098, 0.05880556607209096, 0.05880256329021113]
>       COLLATE_FUNCTION: multicrop_collator 
>       MMAP_MODE: True 
>       COPY_TO_LOCAL_DISK: False
>       COPY_DESTINATION_DIR: /tmp/imagenet1k/ 
>       DROP_LAST: True
>   TRAINER:
>     TRAIN_STEP_NAME: standard_train_step
>   METERS:
>     name: ""
>   MODEL:
>     TRUNK:
>       NAME: resnet
>       RESNETS:
>         DEPTH: 50
>     HEAD:
>       PARAMS: [
>         ["mlp", {"dims": [2048, 2048], "use_relu": True, "skip_last_layer_relu_bn": False}],
>         ["mlp", {"dims": [2048, 128]}],
>       ]
>     TEMP_FROZEN_PARAMS_ITER_MAP: [] 
>     SYNC_BN_CONFIG:
>       CONVERT_BN_TO_SYNC_BN: True  
>       SYNC_BN_TYPE: apex
>       GROUP_SIZE: 8 
>     AMP_PARAMS:
>       USE_AMP: True  
>       AMP_ARGS: {"opt_level": "O1"}  
>   LOSS:
>     name: deepclusterv2_loss
>     deepclusterv2_loss:
>       num_train_samples: 11947
>       BATCHSIZE_PER_REPLICA: 16
>       num_crops: 2
>       temperature: 0.1
>       #num_clusters: [3000, 3000, 3000]
>       num_clusters: [10, 10, 10]
>       kmeans_iters: 10
>       memory_params:
>         crops_for_mb: [0]
>         embedding_dim: 128
>   OPTIMIZER:
>       name: sgd
>       use_larc: True 
>       larc_config: 
>         clip: False
>         trust_coefficient: 0.001
>         eps: 0.00000001
>       weight_decay: 0.000001
>       momentum: 0.9
>       nesterov: False
>       num_epochs: 100
>       regularize_bn: True 
>       regularize_bias: True 
>       param_schedulers:
>         lr: 
>           auto_lr_scaling:
>             auto_scale: true 
>             base_value: 0.3 
>             base_lr_batch_size: 256 
>           name: composite    
>           schedulers:
>             - name: linear    
>               start_value: 0.3
>               end_value: 4.8
>             - name: cosine
>               start_value: 4.8
>               end_value: 0.0000
>           update_interval: step
>           interval_scaling: [rescaled, fixed]
>           lengths: [0.1, 0.9]                 # 100ep FOR EPOCHS VARY 
>   DISTRIBUTED:
>     BACKEND: nccl
>     NUM_NODES: 1    
>     NUM_PROC_PER_NODE: 4 
>     RUN_ID: auto 
>     INIT_METHOD: tcp 
>     NCCL_DEBUG: True
>   TENSORBOARD_SETUP:
>     USE_TENSORBOARD: True
>     LOG_DIR: "/p/home/jusers/chatterjee1/juwels/operation_vissl"
>     EXPERIMENT_LOG_DIR: "tensorboard"
>     FLUSH_EVERY_N_MIN: 5
>     LOG_PARAMS_GRADIENTS: True
>     LOG_PARAMS: True
>     LOG_PARAMS_EVERY_N_ITERS: -1
>   MACHINE:
>     DEVICE: gpu
>   CHECKPOINT:
>     AUTO_RESUME: True
>     DIR: /p/project/deepacf/kiste/DC/checkpoints_evaluate
>     CHECKPOINT_FREQUENCY: 1           
>     CHECKPOINT_ITER_FREQUENCY: 100
>     OVERWRITE_EXISTING: true

**Comment by QuentinDuval** [comment]
Hi @DC95,

The error is coming from the un-synchronisation between the tutorials and the code.

A recent fix (https://github.com/facebookresearch/vissl/commit/bb78ee74b612b07a1a152fb8584cfdee95c41404) introduced a change in the function "extract features" which is correctly impacted in the repo, but not correctly impacted in the tutorial file you mentioned (https://dl.fbaipublicfiles.com/vissl/tutorials/run_distributed_engines.py).

Could you try again with the latest version of VISSL and using the file in the repository instead?
https://github.com/facebookresearch/vissl/blob/master/tools/run_distributed_engines.py

@prigoyal, we need to update the files in tutorials. We also need to have a look at how to keep in sync the files in the tutorial (or re-write the tutorials so that they download the files from the GitHub repository - the unit test do not catch that)

Quentin

**Comment by DC95** [comment]
Hi @QuentinDuval 

Yes, I changed the run_distributed_engines.py as given in the latest repository and the type error is gone. But there is another error that popped up - 

> _File "/p/project/deepacf/kiste/DC/vissl/vissl/trainer/trainer_main.py", line 431, in _extract_split_features
>     "input": torch.cat(sample["data"]).cuda(non_blocking=True),
> **TypeError: expected Tensor as element 0 in argument 0, but got list**_

I checked the function **def extract(self, output_folder: str) -> None:** in **trainer_main.py** and I understand its meant for feature extraction.

My dataset_catalog.json looks like - 
`{"dummy_data_folder": {"train": ["/p/project/deepacf/kiste/DC/dataset/cot_data/train", ""], "val": ["/p/project/deepacf/kiste/DC/dataset/cot_data/test", ""]}}`

Can you suggest, please? (the log file is attached) 
[log.txt](https://github.com/facebookresearch/vissl/files/6828983/log.txt)

The extraction configuration is simple_extract.yaml file as mentioned above.

**Comment by QuentinDuval** [comment]
Hi @DC95,

I was able to reproduce the issue with this configuration file on my side:

```
# @package _global_
engine_name: extract_features
config:
  DATA:
    NUM_DATALOADER_WORKERS: 5
    TRAIN:
      DATA_SOURCES: [disk_folder]
      DATASET_NAMES: [imagenette_160_folder]
      BATCHSIZE_PER_REPLICA: 16
      LABEL_TYPE: sample_index
      TRANSFORMS:
        - name: ImgPilToMultiCrop
          total_num_crops: 2
          size_crops: [224]
          num_crops: [2]
          crop_scales: [[0.08, 1]]
        - name: RandomHorizontalFlip
          p: 0.5
        - name: ImgPilColorDistortion
          strength: 1.0
        - name: ImgPilGaussianBlur
          p: 0.5
          radius_min: 0.1
          radius_max: 2.0
        - name: ToTensor
        - name: Normalize
          mean: [0.004819399821419931, 0.004819357153477664, 0.004819095912873002]
          std: [0.05880587872490098, 0.05880556607209096, 0.05880256329021113]
      COLLATE_FUNCTION: multicrop_collator #
      MMAP_MODE: True
      COPY_TO_LOCAL_DISK: False
      DROP_LAST: False
    TEST:
      DATA_SOURCES: [disk_folder]
      DATASET_NAMES: [imagenette_160_folder]
      BATCHSIZE_PER_REPLICA: 16
      LABEL_TYPE: sample_index
      TRANSFORMS:
        - name: ImgPilToMultiCrop
          total_num_crops: 2
          size_crops: [224]
          num_crops: [2]
          crop_scales: [[0.08, 1]]
        - name: RandomHorizontalFlip
          p: 0.5
        - name: ImgPilColorDistortion
          strength: 1.0
        - name: ImgPilGaussianBlur
          p: 0.5
          radius_min: 0.1
          radius_max: 2.0
        - name: ToTensor
        - name: Normalize
          mean: [0.004819399821419931, 0.004819357153477664, 0.004819095912873002]
          std: [0.05880587872490098, 0.05880556607209096, 0.05880256329021113]
      COLLATE_FUNCTION: multicrop_collator #
      MMAP_MODE: False
      COPY_TO_LOCAL_DISK: False
      DROP_LAST: False
  MODEL:
    TRUNK:
      NAME: resnet
      RESNETS:
        DEPTH: 50
  DISTRIBUTED:
    NUM_NODES: 1
    NUM_PROC_PER_NODE: 2
```

I will debug this and come back to you.

**Comment by QuentinDuval** [comment]
Hi @DC95,

I found what the issue was: the configuration is using the `multicrop_collator` (along with the `ImgPilToMultiCrop`) which are transformations that are meant for SwAV pre-training only.

This configuration will for instance work fine:
- I switched the transform pipeline to a simple center crop, removing the transform and collator of SwAV
- I added a small item of configuration to dump features by big enough chunks (CHUNK_THRESHOLD): you can configure this to lower values depending on the available CPU RAM on your machine

```
# @package _global_
engine_name: extract_features
config:
  DATA:
    NUM_DATALOADER_WORKERS: 5
    TRAIN:
      DATA_SOURCES: [disk_folder]
      DATASET_NAMES: [imagenette_160_folder]
      BATCHSIZE_PER_REPLICA: 16
      LABEL_TYPE: sample_index
      TRANSFORMS:
        - name: Resize
          size: 256
        - name: CenterCrop
          size: 224
        - name: ToTensor
        - name: Normalize
          mean: [0.004819399821419931, 0.004819357153477664, 0.004819095912873002]
          std: [0.05880587872490098, 0.05880556607209096, 0.05880256329021113]
      # COLLATE_FUNCTION: multicrop_collator
    TEST:
      DATA_SOURCES: [disk_folder]
      DATASET_NAMES: [imagenette_160_folder]
      BATCHSIZE_PER_REPLICA: 16
      LABEL_TYPE: sample_index
      TRANSFORMS:
        - name: Resize
          size: 256
        - name: CenterCrop
          size: 224
        - name: ToTensor
        - name: Normalize
          mean: [0.004819399821419931, 0.004819357153477664, 0.004819095912873002]
          std: [0.05880587872490098, 0.05880556607209096, 0.05880256329021113]
      # COLLATE_FUNCTION: multicrop_collator
  MODEL:
    FEATURE_EVAL_SETTINGS:
      EVAL_MODE_ON: True
      FREEZE_TRUNK_ONLY: True
      EXTRACT_TRUNK_FEATURES_ONLY: True
      SHOULD_FLATTEN_FEATS: False
      LINEAR_EVAL_FEAT_POOL_OPS_MAP: [
        ["conv1", ["AvgPool2d", [[10, 10], 10, 4]]],
        ["res2", ["AvgPool2d", [[16, 16], 8, 0]]],
        ["res3", ["AvgPool2d", [[13, 13], 5, 0]]],
        ["res4", ["AvgPool2d", [[8, 8], 3, 0]]],
        ["res5", ["AvgPool2d", [[6, 6], 1, 0]]],
        ["res5avg", ["Identity", []]],
      ]
    TRUNK:
      NAME: resnet
      RESNETS:
        DEPTH: 50
  EXTRACT_FEATURES:
    CHUNK_THRESHOLD: 100_000
  DISTRIBUTED:
    NUM_NODES: 1
    NUM_PROC_PER_NODE: 2
```

Now, if you want to use the SwAV transformations for the extraction of features, we can think about that on our side. In our use cases, we typically do not use transformations during the feature extracts, mostly because we want the features of the original image, but out of curiosity, what is the rationale behind applying transformation for the extract of features (especially having multiple crops)?

I hope this helps :)

**Comment by pranavsinghps1** [comment]
> Hi @DC95,
> 
> The error is coming from the un-synchronisation between the tutorials and the code.
> 
> A recent fix ([bb78ee7](https://github.com/facebookresearch/vissl/commit/bb78ee74b612b07a1a152fb8584cfdee95c41404)) introduced a change in the function "extract features" which is correctly impacted in the repo, but not correctly impacted in the tutorial file you mentioned (https://dl.fbaipublicfiles.com/vissl/tutorials/run_distributed_engines.py). Sorry for that :/
> 
> Could you try again with the latest version of VISSL and using the file in the repository instead?
> https://github.com/facebookresearch/vissl/blob/master/tools/run_distributed_engines.py
> 
> @prigoyal, we need to update the files in tutorials. We also need to have a look at how to keep in sync the files in the tutorial (or re-write the tutorials so that they download the files from the GitHub repository - the unit test do not catch that)
> 
> Quentin

I have addressed a few tutorials related issues #358 , #337, and #347. Let me know if I can be of any assistance in helping you guys out! :)

**Comment by DC95** [comment]
Hi @QuentinDuval, and thank you, @pranavsinghps1, for the assistance offered.

I tried with the above-mentioned changes, and the feature extraction works now. Thank you very much for your assistance.

Please note - We can close this issue as the purpose is solved.
